Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm

```diff
@@ -1288,17 +1298,31 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
     legacyMainFrameProcess().sendWithAsyncReply(Messages::WebPage::UpdateUnderlyingTextVisibilityForTextAnimationID(uuid, visible), WTFMove(completionHandler), webPageIDInMainFrameProcess());
 }
 
-void WebPageProxy::showSelectionForActiveWritingToolsSession()
+void WebPageProxy::didEndPartialIntelligenceTextPonderingAnimationImpl()
 {
-    if (!hasRunningProcess())
+    if (!internals().partialIntelligenceTextPonderingAnimationCount) {
+        ASSERT_NOT_REACHED();
         return;
+    }
 
-    legacyMainFrameProcess().send(Messages::WebPage::ShowSelectionForActiveWritingToolsSession(), webPageIDInMainFrameProcess());
+    internals().partialIntelligenceTextPonderingAnimationCount -= 1;
 }
 
 void WebPageProxy::didEndPartialIntelligenceTextPonderingAnimation(IPC::Connection&)
 {
-    protectedPageClient()->didEndPartialIntelligenceTextPonderingAnimation();
+    didEndPartialIntelligenceTextPonderingAnimationImpl();
+}
+
+void WebPageProxy::showSelectionForActiveWritingToolsSessionIfNeeded()
+{
+    if (!hasRunningProcess())
+        return;
+
+    if (!internals().partialIntelligenceTextPonderingAnimationCount && internals().writingToolsTextReplacementsFinished) {
```

Comment:
After reading this code, I think the state / mutation of these properties might make more sense on `WKWebView`. With a `PageClient` method to expose the result of this condition.